### PR TITLE
Adding IPVS blog post

### PR DESCRIPTION
**Pull request by kbarnard10**
For 5 Days of K8s series.

/assign @natekartchner

### CODE REVIEW

File: content/en/blog/_posts/2018-07-09-IPVS-In-Cluster-Load-Balancing.md [hunk 1]
```diff
@@ -0,0 +1,181 @@
...
+
+In the future, we can implement Service specific scheduler (potentially via annotation), which has higher priority and overwrites the value.
+
+**Parameter: --cleanup-ipvs** Similar to the --cleanup-iptables parameter, if true, cleanup IPVS configuration and IPTables rules that are created in IPVS mode.
+
+**Parameter: --ipvs-sync-period** Maximum interval of how often IPVS rules are refreshed (e.g. '5s', '1m'). Must be greater than 0.
+
+**Parameter: --ipvs-min-sync-period** Minimum interval of how often the IPVS rules are refreshed (e.g. '5s', '1m'). Must be greater than 0.
+
+**Parameter: --ipvs-exclude-cidrs**  A comma-separated list of CIDR's which the IPVS proxier should not touch when cleaning up IPVS rules because IPVS proxier can't distinguish kube-proxy created ipvs rules from user original ipvs rules. If you are using IPVS proxier with your own IPVS rules in the environment, this parameter should be specified, otherwise your original rule will be cleaned.
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-09-IPVS-In-Cluster-Load-Balancing.md]: uppercase `ipvs` -> `IPVS`.
> distinguish kube-proxy created IPVS rules from user original IPVS rules.

File: content/en/blog/_posts/2018-07-09-IPVS-In-Cluster-Load-Balancing.md [hunk 2]
```diff
@@ -0,0 +1,181 @@
...
+
+## Introduction
+
+Per [the Kubernetes 1.11 release blog post ](https://kubernetes.io/blog/2018/06/27/kubernetes-1.11-release-announcement/), we announced that IPVS-Based In-Cluster Service Load Balancing graduates to General Availability. In this blog, we will take you through a deep dive of the feature.
+
+## What Is IPVS?
+
+**IPVS** (**IP Virtual Server**) is built on top of the Netfilter and implements transport-layer load balancing as part of the Linux kernel.
+
+IPVS is incorporated into the LVS(Linux Virtual Server), where it runs on a host and acts as a load balancer in front of a cluster of real servers. IPVS can direct requests for TCP- and UDP-based services to the real servers, and make services of the real servers appear as virtual services on a single IP address. Therefore, IPVS naturally supports Kubernetes Service.
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-09-IPVS-In-Cluster-Load-Balancing.md]: add space before bracket:
`LVS(Linux Virtual Server)` -> `LVS (Linux Virtual Server)`

File: content/en/blog/_posts/2018-07-09-IPVS-In-Cluster-Load-Balancing.md [hunk 3]
```diff
@@ -0,0 +1,181 @@
...
+
+Kube-proxy, the building block of service routing has relied on the battle-hardened iptables to implement the core supported Service types such as ClusterIP and NodePort. However, iptables struggles to scale to tens of thousands of Services because it is designed purely for firewalling purposes and is based on in-kernel rule lists.
+
+Even though Kubernetes already support 5000 nodes in release v1.6, the kube-proxy with iptables is actually a bottleneck to scale the cluster to 5000 nodes. One example is that with NodePort Service in a 5000-node cluster, if we have 2000 services and each services have 10 pods, this will cause at least 20000 iptable records on each worker node, and this can make the kernel pretty busy.
+
+On the other hand, using IPVS-based in-cluster service load balancing can help a lot for such cases. IPVS is specifically designed for load balancing and uses more efficient data structures(hash tables) allowing for almost unlimited scale under the hood.
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-09-IPVS-In-Cluster-Load-Balancing.md]: space before `(hash tables)`.

File: content/en/blog/_posts/2018-07-09-IPVS-In-Cluster-Load-Balancing.md [hunk 4]
```diff
@@ -0,0 +1,181 @@
...
+
+## IPVS-based Kube-proxy
+
+### Parameter Changes
+
+**Parameter: --proxy-mode** In addition to existing userspace and iptables modes, IPVS mode is configured via `--proxy-mode=ipvs`. It implicitly uses IPVS NAT mode for Service port mapping.
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-09-IPVS-In-Cluster-Load-Balancing.md]: possibly lowercase `service`?

File: content/en/blog/_posts/2018-07-09-IPVS-In-Cluster-Load-Balancing.md [hunk 5]
```diff
@@ -0,0 +1,181 @@
...
+
+**Parameter: --ipvs-scheduler**
+
+A new kube-proxy parameter will be added to specify the IPVS load balancing algorithm, with the parameter being `--ipvs-scheduler`. If it’s not configured, then round-robin (rr) is the default value.
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-09-IPVS-In-Cluster-Load-Balancing.md]: re: `will be added`
`--ipvs-scheduler` was fist added in `v1.8.0-beta.0`, 
while the scheduler methods were introduced around `v1.9.0-alpha.2`, if i'm not mistaken.

perhaps reword the sentence to use paste tense?

File: content/en/blog/_posts/2018-07-09-IPVS-In-Cluster-Load-Balancing.md [hunk 6]
```diff
@@ -0,0 +1,181 @@
...
+- rr: round-robin
+- lc: least connection
+- dh: destination hashing
+- sh: source hashing
+- sed: shortest expected delay
+- nq: never queue
+
+In the future, we can implement Service specific scheduler (potentially via annotation), which has higher priority and overwrites the value.
+
+**Parameter: --cleanup-ipvs** Similar to the --cleanup-iptables parameter, if true, cleanup IPVS configuration and IPTables rules that are created in IPVS mode.
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-09-IPVS-In-Cluster-Load-Balancing.md]: style `--cleanup-iptables parameter` with " ` "

File: content/en/blog/_posts/2018-07-09-IPVS-In-Cluster-Load-Balancing.md (hunk 1 above)
Comment (neolit123) [reply content/en/blog/_posts/2018-07-09-IPVS-In-Cluster-Load-Balancing.md]: possibly style `--ipvs-exclude-cidrs`, `--ipvs-min-sync-period`, `--ipvs-sync-period` with " ` "
even if they are in bold text?

File: content/en/blog/_posts/2018-07-09-IPVS-In-Cluster-Load-Balancing.md [hunk 7]
```diff
@@ -0,0 +1,181 @@
...
+
+**Parameter: --ipvs-exclude-cidrs**  A comma-separated list of CIDR's which the IPVS proxier should not touch when cleaning up IPVS rules because IPVS proxier can't distinguish kube-proxy created ipvs rules from user original ipvs rules. If you are using IPVS proxier with your own IPVS rules in the environment, this parameter should be specified, otherwise your original rule will be cleaned.
+
+### Design Considerations
+
+#### IPVS Service Network Topology
+
+When creating a ClusterIP type Service, IPVS proxier will do the following three things:
+
+* Make sure a dummy interface exists in the node, default to kube-ipvs0
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-09-IPVS-In-Cluster-Load-Balancing.md]: `, default to` ->`, defaults to` ?

File: content/en/blog/_posts/2018-07-09-IPVS-In-Cluster-Load-Balancing.md [hunk 8]
```diff
@@ -0,0 +1,181 @@
...
+# ipvsadm -ln
+IP Virtual Server version 1.2.1 (size=4096)
+Prot LocalAddress:Port Scheduler Flags
+  -> RemoteAddress:Port           Forward Weight ActiveConn InActConn     
+TCP  10.102.128.4:3080 rr
+  -> 10.244.0.235:8080            Masq    1      0          0         
+  -> 10.244.1.237:8080            Masq    1      0          0   
+```
+
+Please note that the relationship between a Kubernetes Service and IPVS virtual servers is `1:N`. Consider a Kubernetes Service that has more than one IP addresses, for example, an External IP type Service has two IP addresses(ClusterIP and External IP). Then the IPVS proxier will create 2 IPVS virtual servers - one for Cluster IP and another one for External IP. The relationship between a Kubernetes Endpoint(each IP+Port pair) and an IPVS virtual server is `1:1`.
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-09-IPVS-In-Cluster-Load-Balancing.md]: spaces before `(each IP+Port pair)` and `(ClusterIP and External IP)`.

i would reword:
> Consider a Kubernetes Service that has more than one IP addresses, for example, an External IP type Service has two IP addresses(ClusterIP and External IP).

to:

> For example, consider a Kubernetes Service that has more than one IP address. An External IP type Service has two IP addresses - ClusterIP and External IP.

File: content/en/blog/_posts/2018-07-09-IPVS-In-Cluster-Load-Balancing.md [hunk 9]
```diff
@@ -0,0 +1,181 @@
...
+
+Deleting of a Kubernetes service will trigger deletion of the corresponding IPVS virtual server, IPVS real servers and its IP addresses bound to the dummy interface.
+
+####  Port Mapping
+
+There are three proxy modes in IPVS: NAT (masq), IPIP and DR. Only NAT mode supports port mapping. Kube-proxy leverages NAT mode for port mapping. The following example shows IPVS  mapping Service port  3080 to Pod port 8080.
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-09-IPVS-In-Cluster-Load-Balancing.md]: there is an extra space before `mapping` and ` 3080`.

File: content/en/blog/_posts/2018-07-09-IPVS-In-Cluster-Load-Balancing.md [hunk 10]
```diff
@@ -0,0 +1,181 @@
...
+# ipvsadm -ln
+IP Virtual Server version 1.2.1 (size=4096)
+Prot LocalAddress:Port Scheduler Flags
+  -> RemoteAddress:Port           Forward Weight ActiveConn InActConn
+TCP  10.102.128.4:3080 rr persistent 10800
+```
+
+#### Iptables & Ipset in IPVS Proxier
+
+IPVS is for load balancing, it can't handle other workarounds in kube-proxy, e.g. packet filtering, hairpin-masquerade tricks, SNAT, etc.
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-09-IPVS-In-Cluster-Load-Balancing.md]: `, it can't ` -> `and it can't `

File: content/en/blog/_posts/2018-07-09-IPVS-In-Cluster-Load-Balancing.md [hunk 11]
```diff
@@ -0,0 +1,181 @@
...
+| KUBE-NODE-PORT-LOCAL-TCP       | NodePort type Service TCP port with `externalTrafficPolicy=local` | accept packets to NodePort Service with `externalTrafficPolicy=local` |
+| KUBE-NODE-PORT-UDP             | NodePort type Service UDP port           | masquerade for packets to NodePort(UDP)  |
+| KUBE-NODE-PORT-LOCAL-UDP       | NodePort type service UDP port with `externalTrafficPolicy=local` | accept packets to NodePort Service with `externalTrafficPolicy=local` |
+
+
+In general, for IPVS proxier, the number of iptables rules is static, no matter how many Services/Pods we have.
+
+### Run kube-proxy in IPVS Mode
+
+Currently, local-up scripts, GCE scripts, and kubeadm support switching IPVS proxy mode via exporting environment variables(`KUBE_PROXY_MODE=ipvs`) or specifying flag(`--proxy-mode=ipvs`). Before running IPVS proxier, please ensure IPVS required kernel modules are already installed.
```
Comment (neolit123) [inline content/en/blog/_posts/2018-07-09-IPVS-In-Cluster-Load-Balancing.md]: spaces before `(`.

**Comment by kbarnard10** [comment]
@neolit123 Thanks! Edits made.
